core/rawdb/freezer.go

```diff
@@ -0,0 +1,281 @@
+			time.Sleep(freezerRecheckInterval)
+			continue
+		}
+		head := ReadHeader(nfdb, hash, *number)
+		if head == nil {
+			log.Error("Current fast block unavailable", "number", *number, "hash", hash)
+			time.Sleep(freezerRecheckInterval)
+			continue
+		}
+		// Seems we have data ready to be frozen, process in usable batches
+		limit := *number - freezerBlockGraduation
+		if limit-f.frozen > 30000 {
+			limit = f.frozen + 30000
+		}
+		var (
+			start = time.Now()
+			first = f.frozen
+			last  common.Hash
+		)
+		for f.frozen < limit {
+			// Retrieves all the components of the canonical block
+			hash := ReadCanonicalHash(nfdb, f.frozen)
+			if hash == (common.Hash{}) {
+				log.Error("Canonical hash missing, can't freeze", "number", f.frozen)
+				break
+			}
+			header := ReadHeaderRLP(nfdb, hash, f.frozen)
+			if len(header) == 0 {
+				log.Error("Block header missing, can't freeze", "number", f.frozen, "hash", hash)
+				break
+			}
+			body := ReadBodyRLP(nfdb, hash, f.frozen)
+			if len(body) == 0 {
+				log.Error("Block body missing, can't freeze", "number", f.frozen, "hash", hash)
+				break
+			}
+			receipts := ReadReceiptsRLP(nfdb, hash, f.frozen)
+			if len(receipts) == 0 {
+				log.Error("Block receipts missing, can't freeze", "number", f.frozen, "hash", hash)
+				break
+			}
+			td := ReadTdRLP(nfdb, hash, f.frozen)
+			if len(td) == 0 {
+				log.Error("Total difficulty missing, can't freeze", "number", f.frozen, "hash", hash)
+				break
+			}
+			// Inject all the components into the relevant data tables
+			if err := f.tables["hashes"].Append(f.frozen, hash[:]); err != nil {
+				log.Error("Failed to deep freeze hash", "number", f.frozen, "hash", hash, "err", err)
+				break
+			}
+			if err := f.tables["headers"].Append(f.frozen, header); err != nil {
+				log.Error("Failed to deep freeze header", "number", f.frozen, "hash", hash, "err", err)
+				break
+			}
+			if err := f.tables["bodies"].Append(f.frozen, body); err != nil {
+				log.Error("Failed to deep freeze body", "number", f.frozen, "hash", hash, "err", err)
+				break
+			}
+			if err := f.tables["receipts"].Append(f.frozen, receipts); err != nil {
+				log.Error("Failed to deep freeze receipts", "number", f.frozen, "hash", hash, "err", err)
+				break
+			}
+			if err := f.tables["diffs"].Append(f.frozen, td); err != nil {
+				log.Error("Failed to deep freeze difficulty", "number", f.frozen, "hash", hash, "err", err)
+				break
+			}
+			log.Trace("Deep froze ancient block", "number", f.frozen, "hash", hash)
+			atomic.AddUint64(&f.frozen, 1) // Only modify atomically
+
+			// If it's the last block, save for reporting
+			if f.frozen == limit-1 {
+				last = hash
+			}
+		}
+		// Batch of blocks have been frozen, flush them before wiping from leveldb
+		if err := f.sync(); err != nil {
+			log.Error("Failed to flush frozen tables", "err", err)
+			time.Sleep(freezerRecheckInterval)
+		}
+		// Wipe out all data from the active database
```

Comment:
Whoa! Two lines up, we log the error `Failed to flush frozen tables`. Then we just wait a bit and continue wiping the data? Is that really the Right Thing To Do ?

Reply:
Whoops, `continue`.

Reply:
Actually, I guess that's a `log.Crit`. No way to recover from this.

Reply:
Alternatively we can also exit the freezer goroutine. Then Geth still keeps running with freezing disabled. This might be desirable to avoid crashing if the HDD fills up.

Reply:
Though an error will just blidly go missing if it's in the logs. Perhaps the good thing is actually to indeed crash?

Reply:
Yes I think a crash is called for here.